staging/src/k8s.io/apimachinery/pkg/labels/selector.go

```diff
@@ -867,25 +867,34 @@ func validateLabelValue(k, v string) error {
 	return nil
 }
 
-// SelectorFromSet returns a Selector which will match exactly the given Set. A
-// nil and empty Sets are considered equivalent to Everything().
+// SelectorFromSet is deprecated and unsafe, because it returns
+// a selector that matches everything on parsing errors such as too long values.
+// Use ValidatedSelectorFromSet or SelectorFromValidatedSet instead.
 func SelectorFromSet(ls Set) Selector {
-	if ls == nil || len(ls) == 0 {
+	selector, err := ValidatedSelectorFromSet(ls)
+	if err != nil {
 		return internalSelector{}
 	}
+	return selector
+}
+
+// ValidatedSelectorFromSet returns a Selector which will match exactly the given Set. A
+// nil and empty Sets are considered equivalent to Everything().
+func ValidatedSelectorFromSet(ls Set) (Selector, error) {
+	if ls == nil || len(ls) == 0 {
+		return internalSelector{}, nil
+	}
 	requirements := make([]Requirement, 0, len(ls))
 	for label, value := range ls {
 		r, err := NewRequirement(label, selection.Equals, []string{value})
```

Comment:
Set is a map[string]string; what's an example one that triggers an error here?

Reply:
https://github.com/kubernetes/kubernetes/pull/89747/files#diff-12ebda951d92c7e0f1739fad4346b488R727

Reply:
That's kinda dumb to fail client side construction for. We should just construct the selector and let the server barf.

Reply:
(do doubt I'm the original author that wrote the "dumb" thing ;) )

Reply:
Alternatively, we can make `SelectorFromSet` an alias for `ValidatedSelectorFromSet` which will then result in the selector being constructed and the server rejecting the request

Reply:
(Now that you mentioned it I find that much nicer than the current behavior). It allows us to do the right thing without breaking anyones code)

Reply:
It might be good to do both? It's rather annoying to not have a version that returns an error.

Reply:
Yeah, that's what I meant. Then we don't really need to deprecate it anymore, right?

Reply:
Yeah, better experience. (I'd still mention it in the godoc comment :) )

Reply:
Updated, PTAL

Reply:
From an API perspective its a bit weird to have both `SelectorFromSet` and `SelectorFromValidatedSet` though, I am inclined to remove the latter, WDYT?

Reply:
I probably would have called the latter `SelectorFromSetValidated` but that's just taste. I think it's OK to have a version of the function that can return an error.

Reply:
No, what I mean is: There is now `SelectorFromSet` (changed in this PR) and `SelectorFromValidatedSet` (pre-existing) that do the same thing: Construct an invalid Selector if they get invalid input. As a user of the `labels` package, its unclear which one I should use and whats the difference is (there is none).

So it would IMHO be good to just remove `SelectorFromValidatedSet` and migrate its users to use `SelectorFromSet` which (as of this PR) does the same thing.

WDYT?

Reply:
If they are so similar that one can just call the other, you can do e.g.

```
// SelectorFromValidatedSet is an alias to preserve existing clients.
var SelectorFromValidatedSet = SelectorFromSet
```